dynamiqs/time_tensor.py

```diff
@@ -0,0 +1,257 @@
+        return time_tensor_add(self, other)
+
+    def __radd__(self, other: Tensor | TimeTensor) -> TimeTensor:
+        if not isinstance(other, (Tensor, TimeTensor)):
+            return NotImplemented
+        return self + other
+
+    def __sub__(self, other: Tensor | TimeTensor) -> TimeTensor:
+        if not isinstance(other, (Tensor, TimeTensor)):
+            return NotImplemented
+        return self + (-other)
+
+    def __rsub__(self, other: Tensor | TimeTensor) -> TimeTensor:
+        if not isinstance(other, (Tensor, TimeTensor)):
+            return NotImplemented
+        return other + (-self)
+
+    def __repr__(self) -> str:
+        return f'<{obj_type_str(self)}>'
+
+    def __str__(self) -> str:
+        return self.__repr__()
+
+    def size(self, dim: int) -> int:
+        """Size along a given dimension."""
+        return self.shape[dim]
+
+    def dim(self) -> int:
+        """Get the number of dimensions."""
+        return len(self.shape)
+
+    @property
+    def ndim(self) -> int:
+        """Number of dimensions."""
+        return self.dim()
+
+
+class ConstantTimeTensor(TimeTensor):
+    def __init__(self, tensor: Tensor):
+        self.tensor = tensor
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return self.tensor.dtype
+
+    @property
+    def device(self) -> torch.device:
+        return self.tensor.device
+
+    @property
+    def shape(self) -> torch.Size:
+        return self.tensor.shape
+
+    def __call__(self, t: float) -> Tensor:
+        return self.tensor
+
+    def view(self, *shape: int) -> TimeTensor:
+        return ConstantTimeTensor(self.tensor.view(*shape))
+
+    def adjoint(self) -> TimeTensor:
+        return ConstantTimeTensor(self.tensor.adjoint())
+
+    def __neg__(self) -> TimeTensor:
+        return ConstantTimeTensor(-self.tensor)
+
+    def __mul__(self, other: Number | Tensor) -> TimeTensor:
+        super().__mul__(other)
+        return ConstantTimeTensor(self.tensor * other)
+
+
+class CallableTimeTensor(TimeTensor):
+    def __init__(self, f: callable[[float], Tensor], f0: Tensor):
+        # f0 carries all the transformation on the shape
+        self.f = f
+        self.f0 = f0
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return self.f0.dtype
+
+    @property
+    def device(self) -> torch.device:
+        return self.f0.device
+
+    @property
+    def shape(self) -> torch.Size:
+        return self.f0.shape
+
+    def __call__(self, t: float) -> Tensor:
+        return self.f(t).view(self.shape)
+
+    def view(self, *shape: int) -> TimeTensor:
+        f = self.f
+        f0 = self.f0.view(*shape)
+        return CallableTimeTensor(f, f0)
+
+    @abstractmethod
+    def adjoint(self) -> TimeTensor:
+        f = lambda t: self.f(t).adjoint()
+        f0 = self.f0.adjoint()
+        return CallableTimeTensor(f, f0)
+
+    def __neg__(self) -> TimeTensor:
+        f = lambda t: -self.f(t)
+        f0 = -self.f0
```

Comment:
Why is this necessary ? It doesn't impact the shape

Reply:
I prefer to keep a correct initial value _in case of_ using `f0` for smthg else. Otherwise we should call it `shape`, but it's annoying bc you can't directly apply view operations on `torch.Size` object.